stdlib/path.gr

```diff
@@ -0,0 +1,571 @@
+ * @returns A path corresponding to the parent directory of the given path
+ * 
+ * @since v0.5.5
+ */
+export let parent = (path: Path<a, b>) =>
+  match (path) {
+    (base, _, [_, ...rest]) => (base, Dir, rest),
+    (Rel(upDirs), _, []) => (Rel(upDirs + 1), Dir, []),
+    (Abs(_) as base, _, []) => (base, Dir, []),
+  }: Path<a, Directory>
+
+/**
+ * Returns the base name (named final segment) of a path.
+ * 
+ * @param path: The path to get the name of
+ * @returns A `Some` variant containing the name of the path, or `None` if the path does not have one
+ * 
+ * @example Path.name(Option.unwrap(Path.relativeFile("./dir/file.txt"))) // Some("file.txt")
+ * @example Path.name(Option.unwrap(Path.relativeDir(".."))) // None
+ * 
+ * @since v0.5.5
+ */
+export let name = (path: Path<a, b>) =>
+  match (path) {
+    (_, _, [name, ..._]) => Some(name),
+    _ => None,
+  }
+
+/**
+ * Returns the name of the file path without the extension.
+ * 
+ * @param path: The path to get the stem of
+ * @returns A `Some` variant containing the stem of the file path, or `None` if the path does not have one
+ * 
+ * @since v0.5.5
+ */
+export let stem = (path: Path<a, File>) =>
+  match (path) {
+    (_, _, [name, ..._]) => {
+      match (String.lastIndexOf(".", name)) {
+        Some(dotI) => {
+          let len = String.length(name)
+          if (dotI > 0 && dotI < len - 1) Some(String.slice(0, dotI, name))
+          else Some(name)
+        },
+        None => Some(name),
+      }
+    },
+    _ => None,
+  }
+
+/**
+ * Returns the extension on the name of the file path.
+ * 
+ * @param path: The path to get the extension of
+ * @returns A `Some` variant containing the extension of the file path, or `None` if the path does not have one
+ * 
+ * @since v0.5.5
+ */
+export let extension = (path: Path<a, File>) =>
+  match (path) {
+    (_, _, [name, ..._]) => {
+      match (String.lastIndexOf(".", name)) {
+        Some(dotI) => {
+          let len = String.length(name)
+          if (dotI > 0 && dotI < len - 1) Some(String.slice(dotI, len, name))
+          else None
+        },
+        None => None,
+      }
+    },
+    _ => None,
+  }
+
+/**
+ * Returns the root of the absolute path.
+ * 
+ * @param path: The absolute path to get the root of
+ * @returns The root of the path
+ * 
+ * @since v0.5.5
+ */
+export let root = (path: Path<Absolute, b>) =>
+  match (path) {
+    (Abs(root), _, _) => root,
+    _ => fail "Impossible: malformed absolute path data",
+  }
+
+/**
+ * Converts the given `Path` into a string, using the specified path encoding.
+ * A trailing slash is added to directory paths.
+ * 
+ * @param path: The `Path` to convert to a string
+ * @param encoding: The `PathEncoding` to use to represent the path as a string
+ * @returns A string representing the given path
+ * 
+ * @example Path.toString(myPath, Path.Windows)
+ * 
+ * @since v0.5.5
+ */
+export let toString = (path: Path<a, b>, encoding) => {
```

Comment:
This needs a matching `fromString` because you will often be working with strings that you don't know whether to pass them to one of the 4 parsing functions.